src/modules/storage/TransactionPool.ts

```diff
@@ -0,0 +1,222 @@
+            let tx_hash_data = tx_hash.toBinary(Endian.Little);
+            await this.query(connection, `
+                DELETE FROM tx_pool WHERE \`key\` = ?;
+                DELETE FROM transaction_pool WHERE tx_hash = ?;
+                DELETE FROM tx_input_pool WHERE tx_hash = ?;
+                DELETE FROM tx_output_pool WHERE tx_hash = ?;`
+            , [tx_hash_data, tx_hash_data, tx_hash_data, tx_hash_data]);
+
+            if (rm_double_spent)
+            {
+                let inv_txs = new Set<string>();
+                this.gatherDoubleSpentTXs(key, inv_txs);
+                for (const input of key.inputs)
+                    this.spenders.delete(hashFull(input).toString());
+                for (const inv_tx_hash_string of inv_txs)
+                    await this.remove(connection, new Hash(inv_tx_hash_string), false);
+            }
+            else
+            {
+                let tx_hash_string = tx_hash.toString();
+                for (const input of key.inputs)
+                {
+                    let in_hash_string = hashFull(input).toString();
+                    let set = this.spenders.get(in_hash_string);
+                    if ((set !== undefined) && (set.has(tx_hash_string)))
+                        set.delete(tx_hash_string)
+                }
+            }
+        }
+        else
+        {
+            let rows = await this.query(connection, `SELECT \`val\` FROM tx_pool WHERE \`key\` = ?;`,
+                [key.toBinary(Endian.Little)]);
+            if (rows.length !== 0)
+            {
+                let tx = Transaction.deserialize(SmartBuffer.fromBuffer(rows[0].val));
+                await this.remove(connection, tx, rm_double_spent);
+            }
+        }
+    }
+
+    /**
+     * Load transactions and make the spender list
+     * @param connection MySQL connection
+     */
+    public async loadSpenderList (connection: mysql.Connection)
+    {
+        this.spenders.clear();
+        let rows = await this.query(connection, "SELECT `key`, `val` FROM tx_pool;", []);
+        for (let row of rows)
+        {
+            let tx = Transaction.deserialize(SmartBuffer.fromBuffer(row.tx));
+            await this.updateSpenderList(tx);
+        }
+    }
+
+    /**
+     * Add the given TX to `spenders` list
+     * @param tx the transaction to add
+     */
+    public updateSpenderList (tx: Transaction)
+    {
+        let tx_hash_string = hashFull(tx).toString();
+
+        // insert each input information of the transaction
+        for (const input of tx.inputs)
+        {
+            const in_hash_string = hashFull(input).toString();
+            // Update the spenders list
+            if (!this.spenders.has(in_hash_string))
+                this.spenders.set(in_hash_string, new Set<string>());
+
+            let set = this.spenders.get(in_hash_string);
+            if ((set !== undefined) && (!set.has(tx_hash_string)))
+                set.add(tx_hash_string);
```

Comment:
?

Reply:
Please check it.

Reply:
> Please check it.

```
            let set = this.spenders.get(in_hash_string);
            if  (!set.has(tx_hash_string))
                set.add(tx_hash_string);
```
This results in an error because the `set` can be `undefined`

Reply:
@TrustHenry I changed it to a simpler code. Check it out, please.